Reduction of Dependencies and Deprecation Warnings

mjpcger opened the pull request:
This pull request consists of three separate commits which can be used (partly) independent from each other.

Commit [Product Maintenance](https://github.com/JavaPOSWorkingGroup/javapos-config-loader/commit/4e3fec226f9e35c6f42ed3dadd0e10b4ea1aa802): This commit reduces deprecation warnings. Deprecated constructors have been replaced by simple assignments. It is independent from the other commits.

Commit [XML Handling Extended By Javax Based Variant](https://github.com/JavaPOSWorkingGroup/javapos-config-loader/commit/1c7230efc61cd476516ed63fe734f8bc9d381c4e): This adds an XML registry populator bases on javax instead of xalan/xerces and allows to remove the dependency from the corresponding external java library. It is independent from the other commits.

Commit [Standard Dependency Reduction](https://github.com/JavaPOSWorkingGroup/javapos-config-loader/commit/d79f2a4c98c120f94b04833aad20b3da23c53fab): This modifies the SimpleXMLRegPopulator class to use the new JavaxRegPopulator class instead of the XercesRegPopulator class and therefore removes the dependency from Xalan/Xerces whenever the default in jpos.properties has not been changed. It is independent from the _Product Maintenance_ commit, but for this commit, the commit _XML Handling Extended By Javax Based Variant_ is mandatory.

Commit [New Device Classes Added](https://github.com/JavaPOSWorkingGroup/javapos-config-loader/commit/6536742fe8a488dcf154d36c3a71be8b590cc4c9): This commit was probably already unnecessary due to the changes made for version 1.16 in the meantime.

mjpcger committed:
New Device Classes Added

Device classes added in UPOS 1.16 have been added for attribute category. Keep in mind: These category names do not match the so-called "programmatic names" in the UPOS 1.16.1 specification because the names in the specification contain spaces erroneously - these spaces have been removed here.

mjpcger committed:
Product Maintenance

Assignments of typed values via deprecated constructors replaced by type-casted values.

mjpcger committed:
XML Handling Extended By Javax Based Variant

The new class JavaxRegPopulator extends the config loader with an XML populator that works with Javax instead of Xerces and thus enables JavaPOS to be used without the external modules of the Xalan / Xerces project.

mjpcger committed:
Standard Dependency Reduction

The default SimpleXmlRegPopulator has been changed to use JavaxRegPopulator instead of XercesRegPoputator now. This allows to use JavaPOS without Xalan/Xerces dependencies.

kuniss commented:
@mjpcger, could you check the test please. There are 5 failing...

mjpcger commented:
The problem is the jcl.dtd file that is not where the parser expects. When I perform the test, I get an error like "file C:/Users/...../_ProjectFolder_/jpos/res/jcl.dtd not found". If

- you copy or move the jpos folder from _ProjectFolder_/src/main/resources to _ProjectFolder_ or
- add folder _ProjectFolder_/src/main/resources to the class path

before starting the test, the test would not report more errors as before the changes. Unfortunately, I do not know how to change the test configuration to do this automatically.

mjpcger commented:
Remark: I tried the first action (copy the folder). I would expect that adding the folder to the class path or setting setting the working directory to _ProjectFolder_/src/main/resources will work as well, but I didn't test this.

mjpcger committed:
Fixes and Enhancements

load() has been changed to load file spefified in jpos.properties now.
load(filename has been changed to search file to be loaded in classpath, if not present.
save() saves implicid attributes as well as all other attributes, even if they consists of an empty string.
Populator name has been changed tonique results for all present populators.

mjpcger committed:
Bug Fix New Populator

Errors in JavaxRegPopulator have been fixed.
Restructuring of class implementation.

mjpcger commented:
Rework of new populator class, inclusive bug fixes.

kuniss commented:
Hey, cool! Now the tests passes!
I guess, I will incorporate the 2 outstanding changes in different versions. First, the changes for UPOS 1.16 into a 3.2 version of javapos-config-loader. And second, these changes for getting rid of Xerces into a major version change 4.0. It's worth a major version change...

For 4.0 I would also try to solve the second long outstanding issue #1.

But this all will not happen this year anymore.
So, Merry Christmas and a Happy new year, @mjpcger!

mjpcger committed:
Buf Fixes

Xerces2RegPopularor did not work correctly whenever one of the optional attributes are missing.

mjpcger commented:
BTW: Have made some bug fixes in Xerces2RegPopulator. Did not work correctly whenever one of the attributes marked as #IMPLIED in jcl.dtd were missing.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.parsers.SAXParser;
```
SAXParser is not used. Remove import.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+
+/**
+ * Simple implementation of the JposRegPopulator that loads and saves
+ * the entries in XML using Javax api.
```
Spelling: "API"?!

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+ * created via reflection when it is defined in the jpos.properties as
+ * the jpos.config.regPopulatorClass
+ * @see jpos.util.JposProperties#JPOS_REG_POPULATOR_CLASS_PROP_NAME
+ * @since 1.16
```
javapos-config-loader component has its own versioning which is independent from UnifiedPOS related components here.
I'm planning to put this in a new version 4.0. So you may add this version here.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+     * 1-arg constructor that takes the unique ID
+     *
+     * @param s the unique ID string
+     * @since 1.3 (Washington DC 2001)
```
These references to an ancient meeting should be removed. Not helpful here at all.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+        else
+            url = createURLFromFile( new File( getPopulatorFileName() ) );
+
+        //<temp>
```
What does mean "<temp>" here?
I guess it was just copied over from the original you used as template. I would removed it.

mjpcger replied:
Do you mean "&lt;temp&gt;"? This has really been copied and can be removed.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+            try( FileOutputStream os = new FileOutputStream( "jpos.xml" ) )
+            { save( entries, os ); }
+            catch (Exception e)
+            { throw e; }
```
Throwing the just caught exception is meaningless.
You should either log it or let it out at all.
Logging should only be done if the caller does not log it either...

mjpcger replied:
I only learned try-with-resource from examples and never used it without catch or finally. But really: Removing the catch seems to be ok for the compile, so I'll remove these catch / throw statements.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+        try( FileOutputStream os = new FileOutputStream( fileName ) )
+        { save(entries, os); }
+        catch ( Exception e )
+        { throw e; }
```
Throwing the just caught exception is meaningless.
See above.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+
+
+    /** @return the Tracer object */
+    protected Tracer getTracer() { return tracer; }
```
Not used at all. I'm proposing to remove it.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+     * @return the default XML file name that this populator will save
+     * entries to
+     */
+    protected String getDefaultXmlFileName() { return xmlFileName; }
```
Not used at all. I'm proposing to remove it.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+     */
+    protected String getDefaultXmlFileName() { return xmlFileName; }
+
+    private void save( Enumeration entries, OutputStream outputStream )
```
Even if the overridden `save` methods from the interface are not type safe, we should made at least our internal methods type safe. So, it would be work to use the `Enumeration<JposEntry>` parameter here.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+    protected String getDefaultXmlFileName() { return xmlFileName; }
+
+    private void save( Enumeration entries, OutputStream outputStream )
+            throws Exception
```
It's not a good style in Java to subsume all exception types under the Exception super type.
Better to make them explicit:
`throws ParserConfigurationException, TransformerFactoryConfigurationError, TransformerException`

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+        insertDateSavedComment( document );
+
+        DOMSource source = new DOMSource( document );
+        Transformer transformer = TransformerFactory.newInstance().newTransformer();
```
Out internal analysis tool detected a potential vulnerability for XXE attacks to the XML parser.

There are 3 solutions proposed. But as far as I can assess it, only the following is applicable:
```java
        TransformerFactory factory = TransformerFactory.newInstance();
        factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "file,jar:file");
        factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "file,jar:file");
        Transformer transformer = factory.newTransformer();
```
But it has to be tested!

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+        transformer.transform( source, new StreamResult( outputStream ) );
+    }
+
+    private Document CreateEmptyDocument()
```
Java convention: This should start with a lower case.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+        DOMImplementation domImpl = DocumentBuilderFactory.newInstance().newDocumentBuilder().getDOMImplementation();
+        DocumentType docType = domImpl.createDocumentType( "JposEntries", DTD_DOC_TYPE_VALUE, DTD_FILE_NAME );
+
+        return domImpl.createDocument(null, "JposEntries", docType);
```
Maybe, it is worth to define a constant for "JposEntries" to avoid typos?
E.g. `XML_TAG_JPOSENTRIES`
We should define them in `jpos.config.simple.xml.XmlRegPopulator`, for all JavaPOS XML registry populator implementations!

mjpcger replied:
Good idea. Who shall make the change?

kuniss replied:
You may add them in `XMLRegPopulator` and use it only in the new implementation. The old, Xerces-one we want to removed either.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+    }
+
+
+    private void insertJposEntriesInDoc( Enumeration entries, Document doc )
```
Even if the overridden `save `methods from the interface are not type safe, we should made at least our internal methods type safe. So, it would be work to use the `Enumeration<JposEntry>` parameter here.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+
+            if( JposEntryUtility.isValidJposEntry( jposEntry ) )
+            {
+                Element jposEntryElement = doc.createElement( "JposEntry" );
```
Maybe, it is worth to define a constant for "JposEntry" to avoid typos?
E.g. `XML_TAG_JPOSENTRY`.
We should define them in `jpos.config.simple.xml.XmlRegPopulator`, for all JavaPOS XML registry populator implementations!

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+
+    private void insertJposPropertiesInElement( Document doc, JposEntry jposEntry, Element jposEntryElement )
+    {
+        Element creation = doc.createElement( "creation" );
```
We should define all XML tags here as constants in `jpos.config.simple.xml.XmlRegPopulator`, for all JavaPOS XML registry populator implementations!

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+        }
+    }
+
+    private List<JposEntry.Prop> getSortedList( Iterator props )
```
What about making this implementation simpler and better readable using java lambdas?
E.g.
```java
    private List<JposEntry.Prop> getSortedList( Iterator<JposEntry.Prop> props ) {
    	List<JposEntry.Prop> result = new ArrayList<>();
    	props.forEachRemaining(result::add);

        Collections.sort(result, 
        		(Prop p1, Prop p2) -> p1.getName().compareToIgnoreCase(p2.getName())
        );
        return result;
    }
```

mjpcger replied:
Sorry I did not know Collection.sort, therefore I wrote my own sort. 
For me, it looks a little bit different: 

    private List<JposEntry.Prop> getSortedList(Iterator<JposEntry.Prop> props) {
        ArrayList<JposEntry.Prop> sortedProps = new ArrayList<>();
        props.forEachRemaining(sortedProps::add);
        Collections.sort(sortedProps, new Comparator<JposEntry.Prop>() {
            @Override
            public int compare(JposEntry.Prop o1, JposEntry.Prop o2) {
                return o1.getName().compareToIgnoreCase(o2.getName());
            }
        });
        return sortedProps;
    }
Is it what you mean?

kuniss replied:
Your code is the same as mine but with an explicit implementation of the Single-method-Interface.
For such single-method-interface implementations since Java8 the lambda notation may be used. Both is right. Lambda is almost preferred, however.
You may leave this to me to change to lambda notation during merging if you are not familiar with lambda... 
If so, just resolve the conversation.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+        ArrayList<JposEntry.Prop> sortedProps = new ArrayList<>();
+        while( props.hasNext() )
+        {
+            JposEntry.Prop entry = (JposEntry.Prop) props.next();
```
This cast is not needed.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+
+    private static void addPropElement( Document doc, Element jposEntryElement, JposEntry.Prop prop )
+    {
+        Element more = doc.createElement( "prop" );
```
Naming: Maybe, its better to name it "propElement" instead of "more"?

mjpcger replied:
I'll change it.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+    {
+        String dateString = DateFormat.getInstance().format( new Date( System.currentTimeMillis() ) );
+        String commentString = "Saved by JavaPOS jpos.config/loader (JCL) version " + Version.getVersionString()
+                + " on " + dateString;
```
I propose to rename the comment string to
`Saved by javapos-config-loader (JCL) version`

mjpcger replied:
I'll change it-

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+        jposEntryList.clear();
+        try
+        {
+            parserFactory.newSAXParser().parse( inputStream, saxHandler );
```
Out internal analysis tool detected a potential vulnerability for XXE attacks to the XML parser.

There are 3 solutions proposed. But as far as I can assess it, only the following is applicable:
```java
        try
        {
            SAXParser parser = parserFactory.newSAXParser();
            parser.setProperty(XMLConstants.ACCESS_EXTERNAL_DTD, "file");
            parser.setProperty(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "file,jar:file");
            parser.parse( inputStream, saxHandler );
        }
```

Not tested, however!

mjpcger replied:
I'll change it.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+            tracer.println( "SAX Parser configuration error, msg=" + e.getMessage() );
+            lastLoadException = e;
+        }
+        Iterator entries = jposEntryList.iterator();
```
Type safety: We should make it type safe as `Iterator<JposEntry>`.

mjpcger replied:
I'll change it.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+        Iterator entries = jposEntryList.iterator();
+        while( entries.hasNext() )
+        {
+            JposEntry jposEntry = (JposEntry) entries.next();
```
If made type safe above, the casting can be eliminated here.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+    private class JavaxSaxHandler
+            extends DefaultHandler
+    {
+        private JposEntry CurrentEntry = null;
```
Java conventions: Should be lower case at the beginning.

mjpcger replied:
I'll change it.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+            extends DefaultHandler
+    {
+        private JposEntry CurrentEntry = null;
+        private SAXException TheException = null;
```
Java conventions: Should be lower case at the beginning.

mjpcger replied:
I'll change it.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+                TheException = null;
+                return;
+            }
+            if( qname.equals( "JposEntries" ) )
```
Use XML tag name constants here in this method as discussed earlier.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+                    temp = attrs.getValue( "type" );
+                    try
+                    {
+                        Class type = temp == null ? String.class : Class.forName( "java.lang." + temp );
```
Compiler complains here about a raw type. However, we do not know the type...
But we can rid of the compiler warning by declaring `Class<?>`.

mjpcger replied:
I'll change it.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+
+    protected String xmlFileName = DEFAULT_XML_FILE_NAME;
+
+    private List jposEntryList = new LinkedList();
```
Type Safety: We should make it type safe as it is a local declaration, i.e.
`    private List<JposEntry> jposEntryList = new LinkedList<>();`

kuniss replied:
Forgot one thing.
The original implementor used LinkedList quite often. 
However, we do not sort the list and do not insert. So, does not make sense to use the more sophisticated `LinkedList`. An `ArrayList`will do it too. 
```suggestion
    private List jposEntryList<JposEntry> = new ArrayList<>();
```

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+    // Instance variables
+    //
+
+    protected String xmlFileName = DEFAULT_XML_FILE_NAME;
```
This should be removed and `jpos.config.simple.xml.XmlRegPopulator.DEFAULT_XML_FILE_NAME` should be used instead.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,513 @@
+    // Constants
+    //
+
+    public static final String DTD_FILE_PATH = "jpos/res";
```
This constant and those below should be moved to `jpos.config.simple.xml.XmlRegPopulator` and used in all XML implementations.

mjpcger replied:
Who shall do this?

kuniss reviewed (changes requested):
@mjpcger, I did a review and had some notes added to the source code, especially at _JavaxRegPopulator.java_.

Some general notes upfront:

We should divide your PR in 3
- one with Xerces-less implementation
- one with corrections (mainly _JposEntryUtility.java_ and _Xeces2regPopulator.java_)
- one with the UPOS 1.16 enhancements in _jcl.dtd_ (this may be skipped as PR at all because it is already implemented in the branch https://github.com/JavaPOSWorkingGroup/javapos-config-loader/blob/javapos116/src/main/resources/jpos/res/jcl.dtd)

So, finally 2 PRs, because I would like to make 2 releases of _javapos-config-loader_. One 3.2 with UPOS 1.16 changes and the corrections. And one with 4.0 which has the Xerces-only implementation and Xerces kicked out, but including the UPOS 1.16 changes.

You should also run a Java formatter over the code. It looks sometimes very old fashioned like the original one. Why not give it your style... 😄 

Last, I removed the Xerces dependency and detected that we have another usage of it in _javapos-config-loader/src/main/java/jpos/profile/XercesProfileFactory.java_ as a compile error came up there. This class is used in `jpos.loader.simple.SimpleServiceManager.getProfileFactory()` only. Which itself is called in `jpos.loader.simple.SimpleServiceManager.loadProfile(String)`. This, however, is not called in _javapos-config-loader_ code at all, but is a public method. So, it needs to be implemented.
It instantiates DOM parser:
```java
    private DOMParser domParser = new DOMParser();
```

mjpcger committed:
Really First Changes After Review

Made the changes that are restricted to JavaxRegPopulator. In addition to requester changes, string constant "jpos.xml" has been replaced by DEFAULT_XML_FILE_NAME in method save,

mjpcger commented:
Have made the changes that only affect JavaxRegPoputator. The changes affectin othe classes (definitions for tag and file names) should be made by another person. I'll replace the corresponding string constants in JavaxRegPopulator when they become available.
Remark: Made a push on wrong branch first, hope it does not affect this work flow.

mjpcger committed:
Bug Fix: Unsupported Attributes

Tests show that attributes ACCESS_EXTERNAL_DTD and ACCESS_EXTERNAL_SCHEMA seem not to be supported by TransformerFactory, therefore setting these attributes has been undone.

mjpcger commented:
Tests failed in method save, TransformerFactory seems not to support the attributes set due to review. Has been undone.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,450 @@
+    }
+
+    @Override
+    public void save(Enumeration entries, String fileName)
```
It's good to get rid of this compiler warning at all, as we will not solve it by intention...

```suggestion
    @SuppressWarnings("unchecked")
    @Override
    public void save(Enumeration entries, String fileName)
```

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,450 @@
+
+
+    @Override
+    public void save(Enumeration entries)
```
It's good to get rid of this compiler warning at all, as we will not solve it by intention... 

```suggestion
    @SuppressWarnings("unchecked")
    @Override
    public void save(Enumeration entries)
```

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,450 @@
+        }
+    }
+
+    private void insertJposPropertiesInElement(Document doc, JposEntry jposEntry, Element jposEntryElement) {
```
@mjpcger I'm missing a quite long comment I gave regarding this method. :-( So, I (maybe) repeat this here.

The method `notAttribute` is a predicate with side effect. This is not a good style as it is unexpected. 
I suggest to implement it differently with filters and methods which are intended to have side effects:

```java
    private void insertJposPropertiesInElement( Document doc, JposEntry jposEntry, Element jposEntryElement )
    {
        Element creation = doc.createElement( "creation" );
        Element jpos = doc.createElement( "jpos" );
        Element product = doc.createElement( "product" );
        Element vendor = doc.createElement( "vendor" );

        for( Element tag : new Element[] { creation, vendor, jpos, product } )
            jposEntryElement.appendChild(tag);

        List<JposEntry.Prop> sortedProps = getSortedList( jposEntry.getProps() );
        List<JposEntry.Prop> attributeStored = sortedProps.stream()
        		.filter(prop -> isAttribute(prop)).collect(Collectors.toList());
        List<JposEntry.Prop> propStored = sortedProps.stream()
        		.filter(prop -> !isAttribute(prop)).collect(Collectors.toList());

        for (Prop prop : attributeStored) {
			addPropAsAttribute(prop, creation, jpos, product, vendor);
		}
        
        for (Prop prop : propStored) {
            addPropElement( doc, jposEntryElement, prop );
		}
        
    }
    
    private static final Set<String> PROPS_AS_ATTRIBUTES = new HashSet<>(Arrays.asList(
    		JposEntry.SERVICE_CLASS_PROP_NAME,
    		JposEntry.SI_FACTORY_CLASS_PROP_NAME,
    		JposEntry.VENDOR_NAME_PROP_NAME,
    		JposEntry.VENDOR_URL_PROP_NAME,
    		JposEntry.JPOS_VERSION_PROP_NAME,
    		JposEntry.DEVICE_CATEGORY_PROP_NAME,
    		JposEntry.PRODUCT_NAME_PROP_NAME,
    		JposEntry.PRODUCT_DESCRIPTION_PROP_NAME,
    		JposEntry.PRODUCT_URL_PROP_NAME
    ));
```

The method `addPropAsAttribute` has still to be implemented selecting the appropriate passed XML element to add the attribute accordingly. A switch statement should do it.

mjpcger replied:
I do not understand what happens here. But I think after I added some comments and changed the name of method _notAttribute_ to _cannotSetAttributeOfTag_, it is not too difficult to understand how these methods work together.
Of cause, if you still think this solution is worse than yours, you are free to change it.

kuniss replied:
OK, we let this review conversation unresolved that I can do it after merging your PR.

kuniss replied:
I had implemented this branch https://github.com/JavaPOSWorkingGroup/javapos-config-loader/commits/xerces-replacement-from-pr7/, see commit f1e974356182b972342024915c64a212259600de.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,450 @@
+    // Instance variables
+    //
+
+    private List<JposEntry> jposEntryList = new LinkedList<>();
```
Forgot one thing.
The original implementor used LinkedList quite often. 
However, we do not sort the list and do not insert. So, does not make sense to use the more sophisticated `LinkedList`. An `ArrayList`will do it too. 
```suggestion
    private List jposEntryList<JposEntry> = new ArrayList<>();
```

kuniss replied:
... Sorting took place afterwards, so I guess LinkedList is OK here.

kuniss commented on src/main/java/jpos/config/simple/xml/JavaxRegPopulator.java:
```diff
@@ -0,0 +1,450 @@
+        DOMSource source = new DOMSource(document);
+        TransformerFactory factory = TransformerFactory.newInstance();
+  //      factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "file,jar:file");
+  //      factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "file,jar:file");
```
I guess, referencing `XMLConstants.ACCESS_EXTERNAL_SCHEMA` is an error in the issue documentation of our security tool. 
It probably should be 

```suggestion
        factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_STYLESHEET, "file,jar:file");
```

I run the XML tests with it and it did not fail anymore.

kuniss reviewed (changes requested):
I made a 2nd review and added some points I forgot last time.
I hope, I found also a solution for the unknown XML factory attribute you saw in the tests.

Please check also the older, still open comments in the files view https://github.com/JavaPOSWorkingGroup/javapos-config-loader/pull/7/files#diff-c4ab6d668b3c110c3cd20cb319d68741543bc2086e06be2e86d429261e1ead68.

mjpcger committed:
Bug Fixes and Enhancements

- Constants for path and names of dtd and xsd file and for the DTD document type value moved to XmlRegPopulator.
- Constants for XML tag and attribute values added to XmlRegPopulator.
- "Unchecked" warning suppressed for some methods.
- TransformerFactory attributes for DTD and staylesheet added.
- Comments for method insertJposPropertiesInElement added.
- Method getSortedList removed. Use Collections.sort directly instead.
- Method notAttribute renamed to cannotSetAttributeOfTag for better understanding.
- Comments for method cannotSetAttributeOfTag added.
- LinkedList replaced by ArrayList  where meaningful and vice versa.

kuniss reviewed (changes requested):
@mjpcger I guess, we are through. 😄 

Could you maybe split these changes in 2 PRs and skip the jcl.dtd change completely as I discussed it in https://github.com/JavaPOSWorkingGroup/javapos-config-loader/pull/7#pullrequestreview-1793739631 , such that I can apply them on different branches?

If it is too complicated for you, I can do it by myself in different branches, in the origin project.

mjpcger commented:
Sounds a little bit complicated. If you can do it by yourself, this would be my favorite solution.

kuniss commented:
I have split this into 2 branches, in both I made some more improvements:
- https://github.com/JavaPOSWorkingGroup/javapos-config-loader/commits/xerces-replacement-from-pr7/, additionally
  - added an explicit test case derived from the Xerces registry populator test case
  - [corrected an initialization error when reloading to the same populator](https://github.com/JavaPOSWorkingGroup/javapos-config-loader/commit/881797ba2504cb6f4cb20c48c7383b272cd9ceb5) 
  - [re-implemented method insertJposPropertiesInElement](https://github.com/JavaPOSWorkingGroup/javapos-config-loader/commit/f1e974356182b972342024915c64a212259600de)
  - introduced generics where possible
  - improved code formatting
- https://github.com/JavaPOSWorkingGroup/javapos-config-loader/tree/fixes-from-pr7, additionally
  - introduced generic types where possible
  - get rid of all compiler warnings
  - removed deprecated methods
  - some other code cleanup of that very old code (from 1999)
  - improved code formatting
The intention is to use the fixes-from-pr7 branch may be for 3.2 again when we make adaptations for UnifiedPOS 1.16. At least we have an option then. So, I will merge the fixes-from-pr7 branch first to main and then the xerces-replacement-from-pr7.

As this PR has been resolved in the 2 branches above, I'm going to close this PR and the discussion here. Please feel free to add notes here if I may have missed something.
